src/services/webhdfs/backend.rs

```diff
@@ -0,0 +1,1026 @@
+// Copyright 2022 Datafuse Labs.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+use core::fmt::Debug;
+use std::io::Read;
+
+use async_trait::async_trait;
+use bytes::Buf;
+use http::header::CONTENT_LENGTH;
+use http::header::CONTENT_TYPE;
+use http::response::Parts;
+use http::Request;
+use http::Response;
+use http::StatusCode;
+use log::debug;
+use log::warn;
+use serde::Deserialize;
+
+use super::auth::WebHdfsAuth;
+use super::dir_stream::DirStream;
+use super::error::blocking_parse_error;
+use super::error::blocking_resp_to_str;
+use super::error::parse_error;
+use super::percent_encode_path;
+use crate::raw::apply_wrapper;
+use crate::raw::build_abs_path;
+use crate::raw::input;
+use crate::raw::new_request_build_error;
+use crate::raw::normalize_root;
+use crate::raw::output;
+use crate::raw::parse_into_object_metadata;
+use crate::raw::Accessor;
+use crate::raw::AccessorCapability;
+use crate::raw::AccessorMetadata;
+use crate::raw::AsyncBody;
+use crate::raw::Body;
+use crate::raw::BytesRange;
+use crate::raw::HttpClient;
+use crate::raw::IncomingAsyncBody;
+use crate::raw::ObjectPage;
+use crate::raw::ObjectPager;
+use crate::raw::RpCreate;
+use crate::raw::RpDelete;
+use crate::raw::RpList;
+use crate::raw::RpRead;
+use crate::raw::RpStat;
+use crate::raw::RpWrite;
+use crate::Error;
+use crate::ErrorKind;
+use crate::ObjectMetadata;
+use crate::ObjectMode;
+use crate::OpCreate;
+use crate::OpDelete;
+use crate::OpList;
+use crate::OpRead;
+use crate::OpStat;
+use crate::OpWrite;
+use crate::Result;
+use crate::Scheme;
+
+const WEBHDFS_DEFAULT_ENDPOINT: &str = "http://127.0.0.1:9870";
+
+type AsyncReq = Request<AsyncBody>;
```

Comment:
Please use `Request<AsyncBody>` directly so that we can understand all services' code.